src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ImplementIDisposableCorrectly.cs

```diff
@@ -106,13 +112,494 @@ public abstract class ImplementIDisposableCorrectlyAnalyzer : DiagnosticAnalyzer
+                            // For example see unit tests:
+                            // CSharp_CA1063_IDisposableReimplementation_NoDiagnostic_ImplementingInheritedInterfaceWithNoDisposeReimplementation
+                            // Basic_CA1063_IDisposableReimplementation_Diagnostic_ImplementingInheritedInterfaceWithNoDisposeReimplementation
+                            CheckIDisposableReimplementationRule(type, context, implementsDisposableInBaseType);
+
+                            CheckDisposeSignatureRule(disposeMethod, type, context);
+                            CheckRenameDisposeRule(disposeMethod, type, context);
+
+                            if (!type.IsSealed && type.DeclaredAccessibility != Accessibility.Private)
+                            {
+                                var disposeBoolMethod = FindDisposeBoolMethod(type);
+                                if (disposeBoolMethod != null)
+                                {
+                                    CheckDisposeBoolSignatureRule(disposeBoolMethod, type, context);
+                                }
+                                else
+                                {
+                                    CheckProvideDisposeBoolRule(type, context);
+                                }
+                            }
+                        }
+                    }
+
+                    if (implementsDisposableInBaseType)
+                    {
+                        foreach (var method in type.GetMembers().OfType<IMethodSymbol>())
+                        {
+                            CheckDisposeOverrideRule(method, type, context);
+                        }
+
+                        CheckFinalizeOverrideRule(type, context);
+                    }
+                }
+            }
+
+            private void AnalyzeOperationBlock(OperationBlockAnalysisContext context)
+            {
+                var method = context.OwningSymbol as IMethodSymbol;
+                if (method == null)
+                {
+                    return;
+                }
+
+                var isFinalizerMethod = method.IsFinalizer();
+                var isDisposeMethod = method.Name == DisposeMethodName;
+                if (isFinalizerMethod || isDisposeMethod)
+                {
+                    var type = method.ContainingType;
+                    if (type != null && type.IsType && type.TypeKind == TypeKind.Class &&
+                        !type.IsSealed && type.DeclaredAccessibility != Accessibility.Private)
+                    {
+                        if (ImplementsDisposableDirectly(type))
+                        {
+                            var disposeMethod = FindDisposeMethod(type);
+                            if (disposeMethod != null)
+                            {
+                                if (method == disposeMethod)
+                                {
```

Comment:
Instead of walking all the blocks yourself, you could call context.RegisterOperationAction() with the operations you are interested in - that way the compiler walks the tree once for all analyzers.